src/librustc_lint/types.rs

```diff
@@ -527,6 +527,28 @@ enum FfiResult<'tcx> {
     FfiUnsafe { ty: Ty<'tcx>, reason: String, help: Option<String> },
 }
 
+/// `repr(transparent)` structs can have a single non-ZST field, this function returns that
+/// field.
+pub fn transparent_newtype_field<'a, 'b, 'tcx>(
+    cx: &'a LateContext<'tcx>,
+    variant: &'b ty::VariantDef,
+) -> Option<&'b ty::FieldDef> {
+    for field in &variant.fields {
+        let field_ty = cx.tcx.type_of(field.did);
+        let is_zst = cx
+            .tcx
+            .layout_of(cx.tcx.param_env(field.did).and(field_ty))
```

Comment:
You can make this cheaper by having `let param_env = cx.tcx.param_env(variant.did);` at the start of the function.
Also, if you want to avoid mistakes here (due to using something related where the lint was running), you can take just a `tcx; TyCtxt<'tcx>` instead of a `LateContext`.

Reply:
Fixed!